ramp4-pcar4/storylines-editor
# Implement ToC redesign/refactor, both-language previews

### Related Item(s)
Issues:
#401 
#402 

### Changes
- [FEATURE] Implements complete redesign and refactor of ToC, based on designs given by UX team.
    - Both language configs are displayed simultaneously, removing the need to swap configs. Creating a new slide creates two new configs, in English and French. 
    - New slide item design, reducing clutter. Each title can take up to two lines now, meaning most titles are fully visible immediately, without tooltips.
    - ToC will now smooth-scroll automatically to the new slide when you create one. It will also smooth scroll to a copied slide until it's in the middle of the ToC, to give a little indicator that something happened.
    - UI element redesign (e.g. buttons)
    - Indicators for slides with a missing config (e.g. a slide doesn't have a FR config). Code handles all edge cases (that I could think of) involving a slide missing one of its language configs. Slides without a config can copy the config of its slide neighbour (e.g. FR version of Slide 12, if it has no config, can copy ENG config of slide 12 if it exists) OR create a new blank config.
    - Slides allow for deleting an individual lang config, as long as its the last of its lang configs in the list (e.g. you can delete an ENG config if its the last ENG config, i.e. there are no ENG configs in any slide below it).
    - ToC should no longer overflow the viewport height, requiring two scrollbars to be used (main page + ToC-specific) to scroll it.
    - A separate ToC is created for mobile sizes. When the viewport is small enough, the ToC will be auto-hidden. Clicking the new hamburger menu up top will open a new mobile ToC drawer, featuring larger buttons and more visual dividers to make it better for touch devices. To access an item's tooltips, you simply long-press it to simulate a hover. Dragging slides to reorder is disabled so users can scroll properly.
- [FEATURE] Implements new preview button, allowing user to select which language config they want to preview.
- [REFACTOR] Complete refactor of code involving slides/configs, to handle new ToC.

### Notes
Implementing the new ToC required a change in how the `slides` object worked. It's now called `bothLanguageSlides`, and each contains two Slide objects, one for each config, as so:
```
export interface SlideForBothLanguages {
    en: Slide | undefined;
    fr: Slide | undefined;
}
```

**App layout**: I've modified the layout so the whole app no longer overshoots the page height (and requires a scrollbar); scrolling is now done only for individual panes (sidebar or slide editor+feedback), when they independently overflow. However, my current solution hardcodes the page height based on the existing elements; it works for different viewport sizes, but will break if there's ANY added/removed page elements in the future that increases/decreases the size given for the ToC or the slide editor half of the page. A future app layout refactor should probably find a better way to do this. 

### TODO
- [x] <del>Mobile ToC (this one doesn't shrink)</del>
- [x] <del>Comment main code</del>

### Testing
Steps:
1. Open the product `00000000-0000-0000-0000-000000000000`.
2. View the new ToC. Test each of the new ToC features listed above, in both the English and French versions of the page.
3. Use the new Preview dropdown to preview both the English and French configs. They should both still show even unsaved changes.


- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/ramp4-pcar4/storylines-editor/409)
